ec2/instance.py

```diff
@@ -14,165 +14,203 @@
+            ctx.instance.runtime_properties['public_dns_name']
+            ctx.instance.runtime_properties['public_ip_address']
     """
     ec2_client = connection.EC2ConnectionClient().client()
 
-    instance_id = ctx.instance.runtime_properties['instance_id']
+    instance_id = ctx.instance.runtime_properties['aws_resource_id']
+
+    if utils.get_instance_state(ctx=ctx) == 16:
+        assign_runtime_properties_to_instance(ctx=ctx)
+        ctx.logger.info('Instance {0} is running.'.format(instance_id))
+        return
 
     ctx.logger.info('Starting EC2 Instance.')
-    ctx.logger.debug('Attempting to start EC2 Classic Instance.'
-                     '(Instance id: {0}.)'.format(instance_id))
+    ctx.logger.debug('Attempting to start instance: {0}.)'.format(instance_id))
 
     try:
-        instances = ec2_client.start_instances(instance_id)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('Error. Failed to start EC2 Instance: '
-                                  'API returned: {0}.'.format(e))
-
-    if utils.validate_instance_id(instance_id, ctx=ctx):
-        utils.validate_state(instances[0], INSTANCE_RUNNING,
-                             START_TIMEOUT, CHECK_INTERVAL, ctx=ctx)
+        ec2_client.start_instances(instance_id)
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        if 'does not exist' in e:
+            raise RecoverableError('Waiting for server to be running'
+                                   ' Retrying...',
+                                   retry_after=retry_interval)
+        else:
+            raise NonRecoverableError('Error. Failed to start EC2 Instance: '
+                                      'API returned: {0}.'.format(str(e)))
+
+    if utils.get_instance_state(ctx=ctx) == 16:
+        assign_runtime_properties_to_instance(ctx=ctx)
+        ctx.logger.info('Instance {0} is running.'.format(instance_id))
+    else:
+        raise RecoverableError('Waiting for server to be running'
+                               ' Retrying...',
+                               retry_after=retry_interval)
 
 
 @operation
-def stop(**kwargs):
+def stop(retry_interval, **_):
     """ Stops an existing EC2 instance.
         If already stopped, this does nothing.
+        Requires:
+            ctx.instance.runtime_properties['aws_resource_id']
     """
     ec2_client = connection.EC2ConnectionClient().client()
 
-    instance_id = ctx.instance.runtime_properties['instance_id']
+    instance_id = ctx.instance.runtime_properties.get('aws_resource_id')
 
     ctx.logger.info('Stopping EC2 Instance.')
     ctx.logger.debug('Attempting to stop EC2 Instance.'
                      '(Instance id: {0}.)'.format(instance_id))
 
     try:
-        instances = ec2_client.stop_instances(instance_id)
-    except (EC2ResponseError, BotoServerError) as e:
+        ec2_client.stop_instances(instance_id)
```

Comment:
maybe we should verify throughout the plugin that when we try to perform actions on actioned resources, we don't get an exception.. but rather boto's implementation is idempotent for all actions....